sdk/src/azure/core/az_log.c

```diff
@@ -16,75 +16,51 @@
 {
   _az_log_message_callback = az_log_message_callback;
 }
 
-// _az_LOG_WRITE_engine is a function private to this .c file; it contains the code to handle
-// _az_LOG_SHOULD_WRITE & _az_LOG_WRITE.
-//
-// If log_it is false, then the function returns true or false indicating whether the message
-// should be logged (without actually logging it).
-//
-// If log_it is true, then the function logs the message (if it should) and returns true or
-// false indicating whether it was logged.
-static bool _az_log_write_engine(bool log_it, az_log_classification classification, az_span message)
+void az_log_set_filter_callback(az_log_should_write_fn az_log_should_write_callback)
 {
-  // Copy the volatile fields to local variables so that they don't change within this function
-  az_log_message_fn const callback = _az_log_message_callback;
-  az_log_classification const* classifications = _az_log_classifications;
-
-  if (callback == NULL)
-  {
-    // If no one is listening, don't attempt to log.
-    return false;
-  }
-
-  if (classifications == NULL)
-  {
-    // If the user hasn't registered any classifications, then we log everything.
-    classifications
-        = &classification; // We don't need AZ_LOG_END_OF_LIST to be there, as very first comparison
-                           // is going to succeed and return from the function.
-  }
-
-  for (az_log_classification const* cls = classifications; *cls != AZ_LOG_END_OF_LIST; ++cls)
-  {
-    // If this message's classification is in the customer-provided list, we should log it.
-    if (*cls == classification)
-    {
-      if (log_it)
-      {
-        callback(classification, message);
-      }
-
-      return true;
-    }
-  }
-
-  // This message's classification is not in the customer-provided list; we should not log it.
-  return false;
+  _az_log_should_write_callback = az_log_should_write_callback;
 }
 
 // This function returns whether or not the passed-in message should be logged.
 bool _az_log_should_write(az_log_classification classification)
 {
-  return _az_log_write_engine(false, classification, AZ_SPAN_EMPTY);
+  _az_PRECONDITION(classification > 0);
+
+  // Copy the volatile fields to local variables so that they don't change within this function.
+  az_log_message_fn const message_callback = _az_log_message_callback;
+  az_log_should_write_fn const should_write_callback = _az_log_should_write_callback;
+
+  // If the user hasn't registered a should_write_callback, then we log everything, as long as a
+  // message_callback metho was provided.
+  return (should_write_callback == NULL || should_write_callback(classification))
+      && message_callback != NULL;
 }
 
 // This function attempts to log the passed-in message.
 void _az_log_write(az_log_classification classification, az_span message)
 {
-  (void)_az_log_write_engine(true, classification, message);
+  _az_PRECONDITION(classification > 0);
+  _az_PRECONDITION_VALID_SPAN(message, 0, true);
```

Comment:
I still don't like the `_az_PRECONDITION(classification > 0)`, but I agree that `_az_PRECONDITION_VALID_SPAN(message, 0, true)` is rather useful, only that if we doing that check, I'd say don't allow empty and null spans, because it is likely a bug, there is no point, at this time at least, in such message.